llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -13474,6 +13476,19 @@ static SDValue expandMul(SDNode *N, SelectionDAG &DAG,
     }
   }
 
+  // 2^N - 2^M -> (sub (shl X, C1), (shl X, C2))
```

Comment:
Isn't it handled by `DAGCombiner::visitMUL`?
https://github.com/llvm/llvm-project/blob/e11b17a4ed90e74147594012207fc35a60515944/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp#L4325-L4371

I'd like to introduce ISD::SHXADD node and move all logic into DAGCombiner.

Reply:
BTW, I think it is time to switch to search-based methods. I believe it will address @wangpc-pp's [concern](https://github.com/llvm/llvm-project/pull/88993#discussion_r1568243508) about the cost model :)

References:
> Bernstein, Robert. "Multiplication by Integer Constants." Software—Practice and Experience 16, 7 (July 1986), 641–652.

Reply:
I think we missed the rule in RISCV's `TLI.decomposeMulByConstant`?

Reply:
One stupid question: Why does it can't be signed int 12 in this line?
https://github.com/llvm/llvm-project/blob/5d314353fbec1a15cd8900f466dcdcf2af40e8c9/llvm/lib/Target/RISCV/RISCVISelLowering.cpp#L20626

Reply:
See https://github.com/llvm/llvm-project/pull/88791 for adding ISD::SHL_ADD..  Once that lands, my hope is to common most of this code across at least x86 and RISCV.  

Please see discussion on https://github.com/llvm/llvm-project/pull/87105 with regards to decomposeMulByConsant.

As a meta point, please don't let perfection be the enemy of the good here.  :)